core/src/main/java/cucumber/runtime/Runtime.java

```diff
@@ -81,20 +81,25 @@ public void run() {
 
         final StepDefinitionReporter stepDefinitionReporter = plugins.stepDefinitionReporter();
         runnerSupplier.get().reportStepDefinitions(stepDefinitionReporter);
-
-        final List<Future<?>> executingPickles = new ArrayList<>();
+        
+        final List<PickleEvent> filteredEvents = new ArrayList<>();
         for (CucumberFeature feature : features) {
             for (final PickleEvent pickleEvent : feature.getPickles()) {
                 if (filters.matchesFilters(pickleEvent)) {
-                    executingPickles.add(executor.submit(new Runnable() {
-                        @Override
-                        public void run() {
-                            runnerSupplier.get().runPickle(pickleEvent);
-                        }
-                    }));
+                	filteredEvents.add(pickleEvent);
                 }
             }
         }
+        
+        final List<Future<?>> executingPickles = new ArrayList<>();
+        for(final PickleEvent pickleEvent : runtimeOptions.getOrderType().orderPickleEvents(filteredEvents)) {        	
```

Comment:
Ideally the order is injected into the `Runtime` via the builder. This allows us and other implementations to create `Runtime` instances without passing in CLI based `RuntimeOptions`.

This will in the long run let us separate the parsing of CLI options from executing pickles.

Reply:
For example:

```
        List<PickleEvent> allPickles = new ArrayList<>();
        for (CucumberFeature feature : features) {
            allPickles.addAll(feature.getPickles());
        }

        List<PickleEvent> orderedPickles = pickleOrder.order(allPickles);
        List<PickleEvent> filteredPickles = pickleFilter.filter(orderedPickles);

        final List<Future<?>> executingPickles = new ArrayList<>();
        for (PickleEvent filtredPickle : filteredPickles) {
            if (filters.matchesFilters(pickleEvent)) {
                executingPickles.add(executor.submit(new Runnable() {
                    @Override
                    public void run() {
                        runnerSupplier.get().runPickle(pickleEvent);
                    }
                }));
            }
        }
```

Reply:
To implement the above the existing iteration for the filters will need to be moved to the `Filters` class. Are you suggesting above that the pickles need to be ordered first and then filtered?

Reply:
Yes. It'd be much easier to implement this with a Java8 `Stream` but we don't have those at the moment. So I don't mind too much where the iteration ends up.